File: .skills/openclaw-skills/skills/ogenes/dingtalk-api/scripts/list-user-cc-approvals.ts
// 获取抄送用户的审批列表
// 用法: ts-node scripts/list-user-cc-approvals.ts <userId> [--startTime <timestamp>] [--endTime <timestamp>] [--maxResults <max>] [--nextToken <token>] [--debug]
// 需要设置环境变量 DINGTALK_APP_KEY 和 DINGTALK_APP_SECRET
export {};

const { default: dingtalkWorkflow, ListUserCcProcessInstancesHeaders, ListUserCcProcessInstancesRequest } = require('@alicloud/dingtalk/workflow_1_0');
const { default: dingtalkOauth2_1_0, GetAccessTokenRequest } = require('@alicloud/dingtalk/oauth2_1_0');
const { Config } = require('@alicloud/openapi-client');
const { RuntimeOptions } = require('@alicloud/tea-util');

interface Result {
  success: boolean;
  userId: string;
  instances: any[];
  totalCount: number;
  hasMore: boolean;
  nextToken?: string;
}

interface ErrorResult {
  success: false;
  error: {
    code: string;
    message: string;
  };
}

function createConfig(): any {
  const config = new Config({});
  config.protocol = "https";
  config.regionId = "central";
  return config;
}

async function getAccessToken(appKey: string, appSecret: string): Promise<string> {
  const config = createConfig();
  const client = new dingtalkOauth2_1_0(config);

  const request = new GetAccessTokenRequest({
    appKey: appKey,
    appSecret: appSecret,
  });

  try {
    const response = await client.getAccessToken(request);
    const accessToken = response.body?.accessToken;

    if (!accessToken) {
      throw new Error('获取 access_token 失败：响应中未包含 token');
    }

    return accessToken;
  } catch (err: any) {
    throw new Error(`获取 access_token 失败: ${err.message || err}`);
  }
}

async function listUserCcApprovals(
  accessToken: string,
  userId: string,
  startTime: number | undefined,
  endTime: number | undefined,
  maxResults: number,
  nextToken: string | undefined,
  debug: boolean
): Promise<void> {
  const client = new dingtalkWorkflow(createConfig());

  const headers = new ListUserCcProcessInstancesHeaders({});
  headers.xAcsDingtalkAccessToken = accessToken;

  const request = new ListUserCcProcessInstancesRequest({
    userId: userId,
    startTime: startTime,
    endTime: endTime,
    maxResults: maxResults,
    nextToken: nextToken,
  });

  try {
    const response = await client.listUserCcProcessInstancesWithOptions(
      request,
      headers,
      new RuntimeOptions({})
    );

    if (debug) {
      console.error('\n=== 调试信息 ===');
      console.error('完整响应:', JSON.stringify(response, null, 2));
      console.error('==============\n');
    }

    const instances = response.body?.result?.data || [];
    const result: Result = {
      success: true,
      userId: userId,
      instances: instances,
      totalCount: instances.length,
      hasMore: !!response.body?.result?.nextToken,
      nextToken: response.body?.result?.nextToken,
    };

    console.log(JSON.stringify(result, null, 2));
  } catch (err: any) {
    const errorResult: ErrorResult = {
      success: false,
      error: {
        code: err.code || 'UNKNOWN_ERROR',
        message: err.message || '未知错误',
      }
    };
    console.error(JSON.stringify(errorResult, null, 2));
    process.exit(1);
  }
}

function parseArgs(args: string[]): { userId: string; startTime?: number; endTime?: number; maxResults: number; nextToken?: string; debug: boolean } {
  let userId = '';
  let startTime: number | undefined;
  let endTime: number | undefined;
  let maxResults = 20;
  let nextToken: string | undefined;
  let debug = false;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--debug') {
      debug = true;
    } else if (arg === '--startTime' && i + 1 < args.length) {
      startTime = parseInt(args[++i], 10);
    } else if (arg === '--endTime' && i + 1 < args.length) {
      endTime = parseInt(args[++i], 10);
    } else if (arg === '--maxResults' && i + 1 < args.length) {
      maxResults = parseInt(args[++i], 10);
    } else if (arg === '--nextToken' && i + 1 < args.length) {
      nextToken = args[++i];
    } else if (!arg.startsWith('--') && !userId) {
      userId = arg;
    }
  }

  return { userId, startTime, endTime, maxResults, nextToken, debug };
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const { userId, startTime, endTime, maxResults, nextToken, debug } = parseArgs(args);

  const appKey = process.env.DINGTALK_APP_KEY;
  const appSecret = process.env.DINGTALK_APP_SECRET;

  if (!appKey || !appSecret) {
    console.error(JSON.stringify({
      success: false,
      error: {
        code: 'MISSING_CREDENTIALS',
        message: '缺少钉钉应用凭证，请设置环境变量 DINGTALK_APP_KEY 和 DINGTALK_APP_SECRET',
        usage: 'export DINGTALK_APP_KEY=your-app-key && export DINGTALK_APP_SECRET=your-app-secret && ts-node scripts/list-user-cc-approvals.ts <userId> [--startTime <timestamp>] [--endTime <timestamp>] [--maxResults <max>] [--nextToken <token>] [--debug]'
      }
    }, null, 2));
    process.exit(1);
  }

  if (!userId) {
    console.error(JSON.stringify({
      success: false,
      error: {
        code: 'INVALID_ARGUMENTS',
        message: '参数错误：需要提供用户 ID（userId）',
        usage: 'ts-node scripts/list-user-cc-approvals.ts <userId> [--startTime <timestamp>] [--endTime <timestamp>] [--maxResults <max>] [--nextToken <token>] [--debug]'
      }
    }, null, 2));
    process.exit(1);
  }

  try {
    console.error('正在获取 access_token...');
    const accessToken = await getAccessToken(appKey, appSecret);
    console.error('access_token 获取成功，正在查询抄送用户的审批列表...');

    await listUserCcApprovals(accessToken, userId, startTime, endTime, maxResults, nextToken, debug);
  } catch (err: any) {
    console.error(JSON.stringify({
      success: false,
      error: {
        code: 'AUTH_FAILED',
        message: err.message || '认证失败',
      }
    }, null, 2));
    process.exit(1);
  }
}

main();

File: .skills/openclaw-skills/skills/hioneowner/ding-skills/scripts/list_user_cc_approvals.py
"""查询用户抄送的审批实例

用法: python scripts/list_user_cc_approvals.py "<userId>" [--startTime <ts>] [--endTime <ts>] [--maxResults 20] [--nextToken "xxx"]
"""

import sys
import os
sys.path.insert(0, os.path.dirname(__file__))
from dingtalk_client import get_access_token, api_request, handle_error, output


def parse_args(argv):
    args = {"user_id": None, "start_time": None, "end_time": None, "max_results": 20, "next_token": None}
    i = 1
    while i < len(argv):
        if argv[i] == "--startTime" and i + 1 < len(argv):
            args["start_time"] = int(argv[i + 1])
            i += 2
        elif argv[i] == "--endTime" and i + 1 < len(argv):
            args["end_time"] = int(argv[i + 1])
            i += 2
        elif argv[i] == "--maxResults" and i + 1 < len(argv):
            args["max_results"] = int(argv[i + 1])
            i += 2
        elif argv[i] == "--nextToken" and i + 1 < len(argv):
            args["next_token"] = argv[i + 1]
            i += 2
        elif argv[i] == "--debug":
            i += 1
        elif not argv[i].startswith("--") and args["user_id"] is None:
            args["user_id"] = argv[i]
            i += 1
        else:
            i += 1
    return args


def main():
    args = parse_args(sys.argv)
    if not args["user_id"]:
        output({"success": False, "error": {"code": "INVALID_ARGS",
            "message": "用法: python scripts/list_user_cc_approvals.py \"<userId>\" [--startTime <ts>] [--endTime <ts>] [--maxResults 20]"}})
        sys.exit(1)

    try:
        token = get_access_token()
        print("正在查询用户抄送的审批...", file=sys.stderr)

        body = {"userId": args["user_id"], "maxResults": args["max_results"]}
        if args["start_time"]:
            body["startTime"] = args["start_time"]
        if args["end_time"]:
            body["endTime"] = args["end_time"]
        if args["next_token"]:
            body["nextToken"] = args["next_token"]

        result = api_request("POST", "/workflow/processInstances/userCc", token, json_body=body)
        instances = result.get("result", {}).get("list", [])
        output({
            "success": True,
            "userId": args["user_id"],
            "instances": instances,
            "totalCount": len(instances),
            "hasMore": bool(result.get("result", {}).get("nextToken")),
            "nextToken": result.get("result", {}).get("nextToken"),
        })
    except Exception as e:
        handle_error(e)
        sys.exit(1)


if __name__ == "__main__":
    main()
